### PR TITLE
Final Pull Request to merge the Finished Project

### PR DESCRIPTION
**Pull request by hemnemne**
## Changelog
* 

* **Please check if the PR fulfills these requirements**
- [ x] The commit message follows our guidelines
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ x] Docs have been added / updated (for bug fixes / features)

* **What kind of change does this PR introduce?** (Bug fix, feature, docs update, ...)

* **Does this PR introduce a breaking change?** (What changes might users need to make in their application due to this PR?)

* **Other information**:(does the pr fix some issues? Tag them with #<issue id>)

  Fixes

### CODE REVIEW

File: app/routes/classifications_upload.py [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+import redis
+from flask import render_template, request, flash, redirect
+from rq import Connection, Queue
+from rq.job import Job
+from werkzeug.utils import secure_filename
+
+from app import app
+from app.forms.classification_form_with_upload import ClassificationFormWithUpload
+#from app.utils.filename_validation import validate_filename
```
Comment (maurapintor) [inline app/routes/classifications_upload.py]: Please, remove leftover comments.

File: app/routes/classifications_id.py [hunk 2]
```diff
@@ -11,7 +10,7 @@
 def classifications_id(job_id):
     """Returns the status and the result of the job identified
     by the id specified in the path."""
-    redis_url = Configuration.REDIS_URL
+    redis_url = config.REDIS_URL
```
Comment (maurapintor) [inline app/routes/classifications_id.py]: why was this changed?

File: app/routes/download_plot.py [hunk 3]
```diff
@@ -0,0 +1,82 @@
+# Add the “download results” button to the web interface.
```
Comment (maurapintor) [inline app/routes/download_plot.py]: This comment could be handled better. First, use the triple quotes ```"""``` comments for docstrings at the top of the files. Second, try to make the docstring more informative about the module, rather than explaining what was the assignment.

File: app/routes/download_results.py [hunk 4]
```diff
@@ -0,0 +1,41 @@
+# Add the “download results” button to the web interface.
```
Comment (maurapintor) [inline app/routes/download_results.py]: Please, fix this docstring as well as described for the other module.

File: app/templates/histogram_output.html [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+                <img class="large-front-thumbnail"
+                     src="{{ url_for('static', filename='imagenet_subset/'+image_id) }} "
+                     alt={{ image_id }}/>
+            </div>
+        </div>
+        <div class="col">
+            <div class="card ">
+                 <img class="large-front-thumbnail"
+                      src= {{ hist }}
+                     <!--src="{{ url_for('static', filename= hist_path) }} " -->
```
Comment (maurapintor) [inline app/templates/histogram_output.html]: Leftover comment here as well.

File: app/utils/filename_validation.py [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    When an image filename containing brackets and blank spaces is uploaded,
+    it is saved in the "uploads" folder without round brackets and 
+    with underscores replacing the blank spaces.
+"""
+def rm_string_spaces(string):
+    """
+    It replaces every blank space in the string with an underscore.
+
+    """
+    list_splits = string.split(" ")
```
Comment (maurapintor) [inline app/utils/filename_validation.py]: I think for this you can use the `str.replace()` of the string type.

File: app/utils/filename_validation.py [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        """if list_splits[i] == '(' or list_splits[i] == ')':
+            continue"""
+        if i == len(list_splits) - 1:
+            new_string += list_splits[i]
+        else:
+            new_string += list_splits[i] + '_'
+    return new_string
+
+def rm_string_round_brackets(string):
+    """
```
Comment (maurapintor) [inline app/utils/filename_validation.py:31]: Same here, use `str.replace()`.

File: ml/classification_utils.py [hunk 8]
```diff
@@ -17,9 +17,14 @@
...
 def fetch_image(image_id):
-    """Gets the image from the specified ID. It returns only images
-    downloaded in the folder specified in the configuration object."""
+    """Gets the image from the specified ID. As default behaviour it
+    returns only images downloaded in the folder specified in the
+    configuration object.
+    If the default path doesn't exist the image might have been uploaded by the user."""
     image_path = os.path.join(conf.image_folder_path, image_id)
+
+    if not os.path.exists(image_path):
```
Comment (maurapintor) [inline ml/classification_utils.py:25]: what if the user uploads an image with the same name as one of the images in the image folder? I know this is unlikely, but we cannot exclude it completely.

**Review by maurapintor — COMMENTED** [review]
Overall, it might work. However, it can be improved in some parts for easier readability and maintainability. I left comments where you can improve the code and make it cleaner.

I don't get why you deleted the files:
* `.github/ISSUE_TEMPLATE/feature_request.md`
* `.github/pull_request_template.md`

Please, fix the remaining issues and email me when you are done.

File: app/routes/classifications_upload.py (hunk 1 above)
Comment (hemnemne) [reply app/routes/classifications_upload.py]: will be done with next pull request

File: app/routes/classifications_id.py (hunk 2 above)
Comment (hemnemne) [reply app/routes/classifications_id.py]: I do not know. But thanks for the hint. I just changed it back since it makes no difference. Tested the functionalities and everything still works. Will also be reverted with next pull request

File: app/routes/download_plot.py (hunk 3 above)
Comment (hemnemne) [reply app/routes/download_plot.py]: Done

File: app/routes/download_results.py (hunk 4 above)
Comment (hemnemne) [reply app/routes/download_results.py]: Done

File: app/templates/histogram_output.html (hunk 5 above)
Comment (hemnemne) [reply app/templates/histogram_output.html]: Done

File: app/utils/filename_validation.py (hunk 6 above)
Comment (hemnemne) [reply app/utils/filename_validation.py]: done

File: app/utils/filename_validation.py (hunk 7 above)
Comment (hemnemne) [reply app/utils/filename_validation.py:31]: done

File: ml/classification_utils.py (hunk 8 above)
Comment (hemnemne) [reply ml/classification_utils.py:25]: done

**Comment by hemnemne** [comment]
I just resolved all comment (hopefully). I will close this Pull Request and open a new on ein a second.
